File: src/operator_middleware_plugs/clients/excel_workbook_client.py
"""Microsoft Graph Excel workbook session client.

# Mythic: Excel Workbook Session
# Engineering: ExcelWorkbookSessionClient
"""

from __future__ import annotations

from typing import Any

import httpx

from src.operator_middleware_plugs.clients.graph_client import call_graph


def _workbook_base(*, item_path: str | None = None, item_id: str | None = None) -> str:
    if item_id:
        return f"me/drive/items/{item_id}/workbook"
    raw = (item_path or "/AAIS/exports/aais.xlsx").lstrip("/")
    safe = "".join(c if c.isalnum() or c in "._/-" else "_" for c in raw)[:200]
    return f"me/drive/root:/{safe}:/workbook"


def create_workbook_session(
    token: str | None,
    *,
    item_path: str | None = None,
    item_id: str | None = None,
    persist_changes: bool = True,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    if not token:
        return {
            "ok": False,
            "reason_code": "EXCEL_NEEDS_AUTH",
            "error": "Graph token required for Excel workbook session",
        }
    base = _workbook_base(item_path=item_path, item_id=item_id)
    res = call_graph(
        token,
        f"{base}/createSession",
        method="POST",
        body={"persistChanges": persist_changes},
        transport=transport,
    )
    if not res.get("ok"):
        return res
    data = res.get("data") if isinstance(res.get("data"), dict) else {}
    session_id = str((data or {}).get("id") or "")
    if not session_id and not res.get("simulated"):
        return {
            "ok": False,
            "reason_code": "EXCEL_SESSION_MISSING_ID",
            "error": "createSession returned no id",
            "data": data,
        }
    return {
        **res,
        "reason_code": "EXCEL_SESSION_SIMULATE" if res.get("simulated") else "EXCEL_SESSION_CREATED",
        "session": {"workbookPath": base, "sessionId": session_id or "sim-session"},
    }


def _session_call(
    token: str | None,
    path: str,
    *,
    method: str,
    body: dict[str, Any] | None,
    session_id: str,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    """call_graph does not accept custom headers; use httpx directly for session header."""
    if not token:
        return {"ok": False, "reason_code": "EXCEL_NEEDS_AUTH", "error": "Graph token required"}
    if not token:
        return call_graph(None, path, method=method, body=body, transport=transport)
    try:
        with httpx.Client(timeout=30.0, transport=transport) as client:
            res = client.request(
                method.upper(),
                f"https://graph.microsoft.com/v1.0/{path.lstrip('/')}",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json",
                    "Accept": "application/json",
                    "workbook-session-id": session_id,
                },
                json=body,
            )
        try:
            data: Any = res.json()
        except Exception:
            data = {"raw": res.text[:2000]}
        if res.status_code >= 400:
            return {
                "ok": False,
                "status": res.status_code,
                "data": data,
                "error": f"Graph HTTP {res.status_code}",
                "reason_code": "GRAPH_HTTP_ERROR",
            }
        return {"ok": True, "status": res.status_code, "data": data, "reason_code": "GRAPH_LIVE_OK"}
    except httpx.HTTPError as exc:
        return {"ok": False, "status": 0, "error": str(exc), "reason_code": "GRAPH_NETWORK_ERROR"}


def get_workbook_range(
    token: str | None,
    session: dict[str, str],
    address: str = "A1:B2",
    *,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    path = f"{session['workbookPath']}/worksheets/Sheet1/range(address='{address}')"
    return _session_call(
        token, path, method="GET", body=None, session_id=session["sessionId"], transport=transport
    )


def update_workbook_range(
    token: str | None,
    session: dict[str, str],
    address: str,
    values: list[list[Any]],
    *,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    path = f"{session['workbookPath']}/worksheets/Sheet1/range(address='{address}')"
    return _session_call(
        token,
        path,
        method="PATCH",
        body={"values": values},
        session_id=session["sessionId"],
        transport=transport,
    )


def close_workbook_session(
    token: str | None,
    session: dict[str, str],
    *,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    return _session_call(
        token,
        f"{session['workbookPath']}/closeSession",
        method="POST",
        body={},
        session_id=session["sessionId"],
        transport=transport,
    )


def run_workbook_session_flow(
    token: str | None,
    *,
    item_path: str | None = None,
    item_id: str | None = None,
    address: str = "A1:B2",
    values: list[list[Any]] | None = None,
    transport: httpx.BaseTransport | None = None,
) -> dict[str, Any]:
    steps: list[dict[str, Any]] = []
    created = create_workbook_session(
        token, item_path=item_path, item_id=item_id, transport=transport
    )
    steps.append({"step": "createSession", **created})
    if not created.get("ok") or not created.get("session"):
        return {
            "ok": False,
            "reason_code": created.get("reason_code"),
            "error": created.get("error"),
            "steps": steps,
        }
    session = created["session"]
    vals = values or [["metric", "value"], ["demo", 1]]
    written = update_workbook_range(token, session, address, vals, transport=transport)
    steps.append({"step": "updateRange", **written})
    if not written.get("ok"):
        close_workbook_session(token, session, transport=transport)
        return {
            "ok": False,
            "reason_code": written.get("reason_code"),
            "error": written.get("error"),
            "steps": steps,
            "sessionId": session.get("sessionId"),
        }
    read = get_workbook_range(token, session, address, transport=transport)
    steps.append({"step": "getRange", **read})
    closed = close_workbook_session(token, session, transport=transport)
    steps.append({"step": "closeSession", **closed})
    data = read.get("data") if isinstance(read.get("data"), dict) else {}
    return {
        "ok": bool(read.get("ok") and closed.get("ok")),
        "reason_code": "EXCEL_SESSION_FLOW_OK" if read.get("ok") else read.get("reason_code"),
        "error": None if read.get("ok") else read.get("error"),
        "steps": steps,
        "sessionId": session.get("sessionId"),
        "values": (data or {}).get("values"),
    }

File: aais-middleware/src/provider_adapters/excel_workbook_client.ts
/**
 * Mythic: Excel Workbook Session
 * Engineering: ExcelWorkbookSessionClient — Graph createSession / range R/W / close
 */
import { callGraph, type FetchLike, type GraphCallResult } from "./graph_client.js";

export interface WorkbookSession {
  workbookPath: string;
  sessionId: string;
}

export interface WorkbookSessionOpts {
  fetchImpl?: FetchLike;
  /** Drive item path under me/drive/root: e.g. /AAIS/exports/report.xlsx */
  itemPath?: string;
  /** Or raw item id: me/drive/items/{id}/workbook/... */
  itemId?: string;
}

function workbookBase(opts: WorkbookSessionOpts = {}): string {
  if (opts.itemId) {
    return `me/drive/items/${encodeURIComponent(opts.itemId)}/workbook`;
  }
  const raw = (opts.itemPath || "/AAIS/exports/aais.xlsx").replace(/^\/+/, "");
  const safe = raw.replace(/[^\w./-]+/g, "_").slice(0, 200);
  return `me/drive/root:/${safe}:/workbook`;
}

function sessionHeaders(sessionId?: string): Record<string, string> {
  return sessionId ? { "workbook-session-id": sessionId } : {};
}

/** Create a persistent workbook session (persistChanges=true). */
export async function createWorkbookSession(
  token: string | undefined,
  opts: WorkbookSessionOpts & { persistChanges?: boolean } = {},
): Promise<GraphCallResult & { session?: WorkbookSession }> {
  if (!token) {
    return {
      ok: false,
      status: 401,
      reasonCode: "EXCEL_NEEDS_AUTH",
      error: "Graph token required for Excel workbook session",
    };
  }
  const base = workbookBase(opts);
  const res = await callGraph(
    token,
    `${base}/createSession`,
    "POST",
    { persistChanges: opts.persistChanges !== false },
    { fetchImpl: opts.fetchImpl },
  );
  if (!res.ok) {
    return { ...res, reasonCode: res.reasonCode || "EXCEL_SESSION_CREATE_FAILED" };
  }
  const data = (res.data || {}) as { id?: string };
  const sessionId = String(data.id || "");
  if (!sessionId && !res.simulated) {
    return {
      ok: false,
      status: res.status,
      reasonCode: "EXCEL_SESSION_MISSING_ID",
      error: "createSession returned no id",
      data: res.data,
    };
  }
  return {
    ...res,
    reasonCode: res.simulated ? "EXCEL_SESSION_SIMULATE" : "EXCEL_SESSION_CREATED",
    session: {
      workbookPath: base,
      sessionId: sessionId || "sim-session",
    },
  };
}

export async function getWorkbookRange(
  token: string | undefined,
  session: WorkbookSession,
  address: string,
  opts?: { fetchImpl?: FetchLike },
): Promise<GraphCallResult> {
  if (!token) {
    return {
      ok: false,
      status: 401,
      reasonCode: "EXCEL_NEEDS_AUTH",
      error: "Graph token required",
    };
  }
  const addr = encodeURIComponent(address || "A1:B2");
  // callGraph doesn't pass custom headers yet — embed session via query workaround:
  // Graph requires workbook-session-id header; extend call with header inject via fetchImpl wrap.
  const fetchImpl = wrapSessionFetch(opts?.fetchImpl, session.sessionId);
  return callGraph(
    token,
    `${session.workbookPath}/worksheets/Sheet1/range(address='${addr}')`,
    "GET",
    undefined,
    { fetchImpl },
  );
}

export async function updateWorkbookRange(
  token: string | undefined,
  session: WorkbookSession,
  address: string,
  values: unknown[][],
  opts?: { fetchImpl?: FetchLike },
): Promise<GraphCallResult> {
  if (!token) {
    return {
      ok: false,
      status: 401,
      reasonCode: "EXCEL_NEEDS_AUTH",
      error: "Graph token required",
    };
  }
  const addr = encodeURIComponent(address || "A1");
  const fetchImpl = wrapSessionFetch(opts?.fetchImpl, session.sessionId);
  return callGraph(
    token,
    `${session.workbookPath}/worksheets/Sheet1/range(address='${addr}')`,
    "PATCH",
    { values },
    { fetchImpl },
  );
}

export async function closeWorkbookSession(
  token: string | undefined,
  session: WorkbookSession,
  opts?: { fetchImpl?: FetchLike },
): Promise<GraphCallResult> {
  if (!token) {
    return {
      ok: false,
      status: 401,
      reasonCode: "EXCEL_NEEDS_AUTH",
      error: "Graph token required",
    };
  }
  const fetchImpl = wrapSessionFetch(opts?.fetchImpl, session.sessionId);
  return callGraph(token, `${session.workbookPath}/closeSession`, "POST", {}, { fetchImpl });
}

/**
 * Full governed flow: create → write → read → close.
 * Fail closed without token; evidence-ready result object.
 */
export async function runWorkbookSessionFlow(
  token: string | undefined,
  input: {
    itemPath?: string;
    itemId?: string;
    address?: string;
    values?: unknown[][];
  },
  opts?: { fetchImpl?: FetchLike },
): Promise<{
  ok: boolean;
  reasonCode: string;
  error?: string;
  steps: Record<string, unknown>[];
  sessionId?: string;
  values?: unknown;
}> {
  const steps: Record<string, unknown>[] = [];
  const created = await createWorkbookSession(token, {
    itemPath: input.itemPath,
    itemId: input.itemId,
    fetchImpl: opts?.fetchImpl,
  });
  steps.push({ step: "createSession", ...created, session: created.session });
  if (!created.ok || !created.session) {
    return {
      ok: false,
      reasonCode: created.reasonCode,
      error: created.error,
      steps,
    };
  }
  const session = created.session;
  const address = input.address || "A1:B2";
  const values = input.values || [
    ["metric", "value"],
    ["demo", 1],
  ];
  const written = await updateWorkbookRange(token, session, address, values, {
    fetchImpl: opts?.fetchImpl,
  });
  steps.push({ step: "updateRange", ...written });
  if (!written.ok) {
    await closeWorkbookSession(token, session, { fetchImpl: opts?.fetchImpl });
    return {
      ok: false,
      reasonCode: written.reasonCode,
      error: written.error,
      steps,
      sessionId: session.sessionId,
    };
  }
  const read = await getWorkbookRange(token, session, address, { fetchImpl: opts?.fetchImpl });
  steps.push({ step: "getRange", ...read });
  const closed = await closeWorkbookSession(token, session, { fetchImpl: opts?.fetchImpl });
  steps.push({ step: "closeSession", ...closed });
  return {
    ok: read.ok && closed.ok,
    reasonCode: read.ok ? "EXCEL_SESSION_FLOW_OK" : read.reasonCode,
    error: read.ok ? undefined : read.error,
    steps,
    sessionId: session.sessionId,
    values: (read.data as { values?: unknown })?.values,
  };
}

function wrapSessionFetch(
  base: FetchLike | undefined,
  sessionId: string,
): FetchLike {
  const inner = base ?? (globalThis.fetch as FetchLike);
  return async (input, init) => {
    const headers = { ...(init?.headers || {}), "workbook-session-id": sessionId };
    return inner(input, { ...init, headers });
  };
}

/** @deprecated use runWorkbookSessionFlow */
export const graphWorkbookStub = async (
  token: string | undefined,
  name: string,
  opts?: { fetchImpl?: FetchLike },
): Promise<GraphCallResult> => {
  const flow = await runWorkbookSessionFlow(
    token,
    { itemPath: `/AAIS/exports/${name}.xlsx` },
    opts,
  );
  return {
    ok: flow.ok,
    status: flow.ok ? 200 : 400,
    reasonCode: flow.reasonCode,
    error: flow.error,
    data: flow,
  };
};
